cmd/doctor.go

```diff
@@ -0,0 +1,577 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juicedata/juicefs/pkg/utils"
+
+	"github.com/urfave/cli/v2"
+)
+
+var defaultOutDir = path.Join(".", "doctor")
+
+const (
+	defaultStatsTime   = 5
+	defaultTraceTime   = 5
+	defaultProfileTime = 30
+)
+
+func cmdDoctor() *cli.Command {
+	return &cli.Command{
+		Name:      "doctor",
+		Action:    doctor,
+		Category:  "INSPECTOR",
+		ArgsUsage: "MOUNTPOINT",
+		Usage:     "Collect and display system static and runtime information",
+		Description: `
+It collects and displays information from multiple dimensions such as the running environment and system logs, etc.
+
+Examples:
+$ juicefs doctor /mnt/jfs
+
+# Result will be output to /var/log/
+$ juicefs doctor --out-dir=/var/log /mnt/jfs
+
+# Get the last up to 1000 log entries
+$ juicefs doctor --out-dir=/var/log --collect-log --limit=1000 /mnt/jfs
+
+# Get pprof information
+$ juicefs doctor --out-dir=/var/log --collect-log --limit=1000 --collect-pprof /mnt/jfs
+`,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:  "out-dir",
+				Value: defaultOutDir,
+				Usage: "the output directory of the result file",
+			},
+			&cli.Uint64Flag{
+				Name:  "stats-sec",
+				Value: defaultStatsTime,
+				Usage: "stats sampling duration",
+			},
+			&cli.BoolFlag{
+				Name:  "collect-log",
+				Usage: "enable log collection",
+			},
+			&cli.Uint64Flag{
+				Name:  "limit",
+				Usage: "the number of last entries to be collected",
+			},
+			&cli.BoolFlag{
+				Name:  "collect-pprof",
+				Usage: "enable pprof collection",
+			},
+			&cli.Uint64Flag{
+				Name:  "trace-sec",
+				Value: defaultTraceTime,
+				Usage: "trace sampling duration",
+			},
+			&cli.Uint64Flag{
+				Name:  "profile-sec",
+				Value: defaultProfileTime,
+				Usage: "profile sampling duration",
+			},
+		},
+	}
+}
+
+func copyVolumeConfWindows(srcPath, destPath string) error {
+	srcFile, err := os.Open(srcPath)
+	if err != nil {
+		return err
+	}
+	defer closeFile(srcFile)
+	destFile, err := os.Create(destPath)
+	if err != nil {
+		return err
+	}
+	defer closeFile(destFile)
+	if _, err := io.Copy(destFile, srcFile); err != nil {
+		return err
+	}
+	return nil
+}
+
+func copyConfigFile(srcPath, destPath string, rootPrivileges bool) error {
+	if runtime.GOOS == "windows" {
+		return copyVolumeConfWindows(srcPath, destPath)
+	}
+
+	var copyArgs []string
+	if rootPrivileges {
+		copyArgs = append(copyArgs, "sudo")
+	}
+	copyArgs = append(copyArgs, "/bin/sh", "-c", fmt.Sprintf("cat %s > %s", srcPath, destPath))
```

Comment:
just call `cp` ?

Reply:
`cp` will have permission issues when compressing the result. cp does not change file permissions when mounted via root. `cat` implements copying while changing file permissions.